File: app/src/main/kotlin/tech/mmarca/openvitals/ui/components/PeriodComparisonStat.kt
package tech.mmarca.openvitals.ui.components

import androidx.compose.material.icons.automirrored.outlined.TrendingDown
import androidx.compose.material.icons.automirrored.outlined.TrendingFlat
import androidx.compose.material.icons.automirrored.outlined.TrendingUp
import androidx.compose.material.icons.Icons
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import tech.mmarca.openvitals.R
import tech.mmarca.openvitals.domain.insights.PeriodComparison
import tech.mmarca.openvitals.domain.insights.PeriodComparisonDirection
import tech.mmarca.openvitals.core.period.TimeRange
import tech.mmarca.openvitals.core.presentation.DisplayValue
import tech.mmarca.openvitals.core.presentation.UnitFormatter
import kotlin.math.abs
import kotlin.math.roundToInt

@Composable
fun previousPeriodInsightStat(
    comparison: PeriodComparison,
    selectedRange: TimeRange,
    unitFormatter: UnitFormatter,
    valueFormatter: @Composable (Double) -> DisplayValue,
    accentColor: Color,
): InsightStat {
    val title = stringResource(
        when (selectedRange) {
            TimeRange.DAY -> R.string.stat_vs_previous_day
            TimeRange.WEEK -> R.string.stat_vs_previous_week
            TimeRange.MONTH -> R.string.stat_vs_previous_month
            TimeRange.YEAR -> R.string.stat_vs_previous_year
        }
    )
    val display = comparison.percentChange
        ?.let { percent ->
            DisplayValue(
                value = signedValue(
                    value = unitFormatter.count(abs(percent.roundToInt())),
                    direction = comparison.direction,
                ),
                unit = stringResource(R.string.unit_percent_symbol),
            )
        }
        ?: valueFormatter(comparison.absoluteChange).let { absolute ->
            DisplayValue(
                value = signedValue(absolute.value, comparison.direction),
                unit = absolute.unit,
            )
        }

    return InsightStat(
        title = title,
        value = display.value,
        unit = display.unit,
        icon = when (comparison.direction) {
            PeriodComparisonDirection.UP -> Icons.AutoMirrored.Outlined.TrendingUp
            PeriodComparisonDirection.DOWN -> Icons.AutoMirrored.Outlined.TrendingDown
            PeriodComparisonDirection.SAME -> Icons.AutoMirrored.Outlined.TrendingFlat
        },
        accentColor = accentColor,
    )
}

private fun signedValue(value: String, direction: PeriodComparisonDirection): String =
    when (direction) {
        PeriodComparisonDirection.UP -> "+$value"
        PeriodComparisonDirection.DOWN -> "-$value"
        PeriodComparisonDirection.SAME -> value
    }

File: lib/ui/components/period_comparison_stat.dart
import 'package:flutter/material.dart';

import '../../core/period/time_range.dart';
import '../../core/presentation/display_value.dart';
import '../../core/presentation/unit_formatter.dart';
import '../../domain/insights/period_comparison.dart';
import '../../l10n/app_localizations.dart';
import 'insight_cards.dart';

/// Port of the Kotlin `PeriodComparisonStat.kt`.

/// Kotlin `signedValue(value, direction)`.
String _signedValue(String value, PeriodComparisonDirection direction) =>
    switch (direction) {
      PeriodComparisonDirection.up => '+$value',
      PeriodComparisonDirection.down => '-$value',
      PeriodComparisonDirection.same => value,
    };

IconData _directionIcon(PeriodComparisonDirection direction) =>
    switch (direction) {
      PeriodComparisonDirection.up => Icons.trending_up,
      PeriodComparisonDirection.down => Icons.trending_down,
      PeriodComparisonDirection.same => Icons.trending_flat,
    };

/// Kotlin `previousPeriodInsightStat`: "Vs previous week, +12 %".
///
/// Falls back to the absolute change when the previous period was ~zero and a
/// percentage would be meaningless (or infinite).
InsightStat previousPeriodInsightStat({
  required PeriodComparison comparison,
  required TimeRange selectedRange,
  required UnitFormatter unitFormatter,
  required DisplayValue Function(double value) valueFormatter,
  required Color accentColor,
  required AppLocalizations l10n,
}) {
  final title = switch (selectedRange) {
    TimeRange.day => l10n.statVsPreviousDay,
    TimeRange.week => l10n.statVsPreviousWeek,
    TimeRange.month => l10n.statVsPreviousMonth,
    TimeRange.year => l10n.statVsPreviousYear,
  };

  final percent = comparison.percentChange;
  final display = percent != null
      ? DisplayValue(
          _signedValue(
            unitFormatter.count(percent.abs().round()),
            comparison.direction,
          ),
          l10n.unitPercentSymbol,
        )
      : () {
          final absolute = valueFormatter(comparison.absoluteChange);
          return DisplayValue(
            _signedValue(absolute.value, comparison.direction),
            absolute.unit,
          );
        }();

  return InsightStat(
    title: title,
    value: display.value,
    unit: display.unit,
    icon: _directionIcon(comparison.direction),
    accentColor: accentColor,
  );
}
